File: internal/handlers/proposal_handler.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ProposalHandler struct {
	db *database.Database
}

func NewProposalHandler(db *database.Database) *ProposalHandler {
	return &ProposalHandler{db: db}
}

type CreateProposalRequest struct {
	ReceiverID string `json:"receiver_id" binding:"required"`
	Title      string `json:"title" binding:"required"`
	Content    string `json:"content" binding:"required"`
}

func (h *ProposalHandler) CreateProposal(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req CreateProposalRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	receiverID, err := primitive.ObjectIDFromHex(req.ReceiverID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid receiver ID"})
		return
	}

	proposal := models.Proposal{
		ID:         primitive.NewObjectID(),
		SenderID:   userIDObj,
		ReceiverID: receiverID,
		Title:      req.Title,
		Content:    req.Content,
		Status:     "pending",
		CreatedAt:  time.Now(),
		UpdatedAt:  time.Now(),
	}

	_, err = h.db.MongoDB.Collection("proposals").InsertOne(context.Background(), proposal)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create proposal"})
		return
	}

	c.JSON(http.StatusCreated, proposal)
}

func (h *ProposalHandler) GetProposals(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	// Get proposals where user is sender or receiver
	cursor, err := h.db.MongoDB.Collection("proposals").Find(
		context.Background(),
		bson.M{
			"$or": []bson.M{
				{"sender_id": userIDObj},
				{"receiver_id": userIDObj},
			},
		},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch proposals"})
		return
	}
	defer cursor.Close(context.Background())

	var proposals []models.Proposal
	if err := cursor.All(context.Background(), &proposals); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode proposals"})
		return
	}

	c.JSON(http.StatusOK, proposals)
}

func (h *ProposalHandler) AcceptProposal(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	proposalIDStr := c.Param("proposal_id")
	proposalID, err := primitive.ObjectIDFromHex(proposalIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid proposal ID"})
		return
	}

	// Verify user is the receiver
	var proposal models.Proposal
	err = h.db.MongoDB.Collection("proposals").FindOne(
		context.Background(),
		bson.M{"_id": proposalID, "receiver_id": userIDObj},
	).Decode(&proposal)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Proposal not found"})
		return
	}

	_, err = h.db.MongoDB.Collection("proposals").UpdateOne(
		context.Background(),
		bson.M{"_id": proposalID},
		bson.M{"$set": bson.M{
			"status":     "accepted",
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to accept proposal"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Proposal accepted"})
}

func (h *ProposalHandler) RejectProposal(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	proposalIDStr := c.Param("proposal_id")
	proposalID, err := primitive.ObjectIDFromHex(proposalIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid proposal ID"})
		return
	}

	// Verify user is the receiver
	var proposal models.Proposal
	err = h.db.MongoDB.Collection("proposals").FindOne(
		context.Background(),
		bson.M{"_id": proposalID, "receiver_id": userIDObj},
	).Decode(&proposal)

	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Proposal not found"})
		return
	}

	_, err = h.db.MongoDB.Collection("proposals").UpdateOne(
		context.Background(),
		bson.M{"_id": proposalID},
		bson.M{"$set": bson.M{
			"status":     "rejected",
			"updated_at": time.Now(),
		}},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to reject proposal"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Proposal rejected"})
}

File: src/handlers/proposalHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { getDB } = require('../database');

// ─── Handlers ─────────────────────────────────────────────────────────────────

async function createProposal(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const { receiver_id, message, is_anonymous } = req.body || {};

    if (!receiver_id) return res.status(400).json({ error: 'receiver_id is required' });

    const now = new Date();
    const proposalDoc = {
      sender_id: userId,
      receiver_id: receiver_id.toString(),
      message: message || '',
      is_anonymous: is_anonymous || false,
      status: 'pending',
      created_at: now,
      updated_at: now,
    };

    const insertResult = await db.collection('proposals').insertOne(proposalDoc);
    proposalDoc._id = insertResult.insertedId;

    return res.status(201).json({ ...proposalDoc, id: insertResult.insertedId.toString() });
  } catch (err) {
    console.error('createProposal error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function getProposals(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;

    const proposals = await db.collection('proposals').find({
      $or: [{ receiver_id: userId }, { sender_id: userId }],
    }).sort({ created_at: -1 }).toArray();

    return res.json(proposals.map(p => ({ ...p, id: p._id.toString() })));
  } catch (err) {
    console.error('getProposals error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function acceptProposal(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const proposalIdStr = req.params.proposal_id;

    let proposalObjId;
    try {
      proposalObjId = new ObjectId(proposalIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid proposal ID' });
    }

    const proposal = await db.collection('proposals').findOne({ _id: proposalObjId });
    if (!proposal) return res.status(404).json({ error: 'Proposal not found' });

    if (proposal.receiver_id !== userId) {
      return res.status(403).json({ error: 'Only the receiver can accept this proposal' });
    }

    const now = new Date();
    await db.collection('proposals').updateOne(
      { _id: proposalObjId },
      { $set: { status: 'accepted', updated_at: now } }
    );

    // Create direct chat between sender and receiver
    const members = [proposal.sender_id, proposal.receiver_id];
    const chatDoc = {
      type: 'direct',
      members,
      created_at: now,
      updated_at: now,
    };

    if (proposal.is_anonymous) {
      chatDoc.anonymous_from_user_id = proposal.sender_id;
    }

    const chatInsert = await db.collection('chats').insertOne(chatDoc);
    const chatIdStr = chatInsert.insertedId.toString();

    return res.json({
      message: 'Proposal accepted',
      chat_id: chatIdStr,
    });
  } catch (err) {
    console.error('acceptProposal error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function rejectProposal(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const proposalIdStr = req.params.proposal_id;

    let proposalObjId;
    try {
      proposalObjId = new ObjectId(proposalIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid proposal ID' });
    }

    const proposal = await db.collection('proposals').findOne({ _id: proposalObjId });
    if (!proposal) return res.status(404).json({ error: 'Proposal not found' });

    if (proposal.receiver_id !== userId) {
      return res.status(403).json({ error: 'Only the receiver can reject this proposal' });
    }

    await db.collection('proposals').updateOne(
      { _id: proposalObjId },
      { $set: { status: 'rejected', updated_at: new Date() } }
    );

    return res.json({ message: 'Proposal rejected' });
  } catch (err) {
    console.error('rejectProposal error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function deleteProposal(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const proposalIdStr = req.params.proposal_id;

    let proposalObjId;
    try {
      proposalObjId = new ObjectId(proposalIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid proposal ID' });
    }

    const deleteResult = await db.collection('proposals').deleteOne({
      _id: proposalObjId,
      sender_id: userId,
    });

    if (deleteResult.deletedCount === 0) {
      return res.status(404).json({ error: 'Proposal not found or not owned by you' });
    }

    return res.json({ message: 'Proposal deleted' });
  } catch (err) {
    console.error('deleteProposal error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

module.exports = {
  createProposal,
  getProposals,
  acceptProposal,
  rejectProposal,
  deleteProposal,
};
